File: Problems/Algorithms/523. Continuous Subarray Sum/ContinuousSubarraySum.java
class Solution {
    public boolean checkSubarraySum(int[] nums, int k) {
        if (nums == null || nums.length < 2) {
            return false;
        }

        Map<Integer, Integer> map = new HashMap<>();
        int[] prefix = new int[nums.length + 1];
        for (int i = 1; i <= nums.length; i++) {
            prefix[i] = prefix[i - 1] + nums[i - 1];
            int rest = prefix[i] % k;
            if (i >= 2 && rest == 0) {
                return true;
            }
            if (map.containsKey(rest)) {
                if (i - map.get(rest) > 1) {
                    return true;
                }
            } else {
                map.put(rest, i);
            }
        }
        return false;
    }
}

File: Problems/Algorithms/523. Continuous Subarray Sum/continuous_subarray_sum.py
#!/usr/bin/env python3

from typing import List


class Solution:
    def checkSubarraySum(self, nums: List[int], k: int) -> bool:
        prefix_mod = 0
        mod_seen = {0: -1}

        for i in range(len(nums)):
            prefix_mod = (prefix_mod + nums[i]) % k

            if prefix_mod in mod_seen:
                if i - mod_seen[prefix_mod] > 1:
                    return True
            else:
                mod_seen[prefix_mod] = i

        return False
